File: problems/0323-number-of-connected-components-in-an-undirected-graph/typescript/solution.ts
// Number of Connected Components — https://leetcode.com/problems/number-of-connected-components-in-an-undirected-graph/
export function countComponents(n: number, edges: number[][]): number {
  const parent = Array.from({ length: n }, (_, i) => i);
  const find = (x: number): number => {
    while (parent[x] !== x) {
      parent[x] = parent[parent[x]];
      x = parent[x];
    }
    return x;
  };
  let count = n;
  for (const [u, v] of edges) {
    const ru = find(u);
    const rv = find(v);
    if (ru !== rv) {
      parent[ru] = rv;
      count--;
    }
  }
  return count;
}

File: problems/0323-number-of-connected-components-in-an-undirected-graph/python/solution.py
def countComponents(n, edges):
    """Number of Connected Components — https://leetcode.com/problems/number-of-connected-components-in-an-undirected-graph/"""
    parent = list(range(n))

    def find(x):
        while parent[x] != x:
            parent[x] = parent[parent[x]]
            x = parent[x]
        return x

    count = n
    for u, v in edges:
        ru, rv = find(u), find(v)
        if ru != rv:
            parent[ru] = rv
            count -= 1
    return count
